### PR TITLE
move mv-async-close from 2.0 branch to 1.4.4

### PR DESCRIPTION
**Pull request by matthewvon**
Moving recent branch against 2.0 to the 1.4 tree.  Notes are here:  https://github.com/basho/leveldb/wiki/mv-async-close

### CODE REVIEW

File: util/env_posix.cc [hunk 1]
```diff
@@ -356,12 +317,36 @@ class PosixMmapFile : public WritableFile {
...
+    // hard code
+    if (!is_async_)
+    {
+        int ret_val;
+
+        ret_val=ftruncate(fd_, file_length);
+        if (0!=ret_val)
+            syslog(LOG_ERR,"Close ftruncate failed [%d, %m]", errno);
+
+        ret_val=close(fd_);
```
Comment (bookshelfdave) [inline util/env_posix.cc:338]: you don't use the ret_val from `close`

File: util/env_posix.cc [hunk 2]
```diff
@@ -400,6 +385,30 @@ class PosixMmapFile : public WritableFile {
...
+          int ret_val;
+
+          ret_val=dec_and_fetch(Count);
+          if (0==ret_val)
+          {
+              ret_val=ftruncate(File, *(Count +1));
+              if (0!=ret_val)
+                  syslog(LOG_ERR,"ReleaseRef ftruncate failed [%d, %m]", errno);
+
+              ret_val=close(File);
```
Comment (bookshelfdave) [inline util/env_posix.cc:405]: same thing, you don't use the ret_val from `close`

File: util/env_posix.cc [hunk 3]
```diff
@@ -652,6 +662,9 @@ class PosixEnv : public Env {
     }
     gFileLocks.Remove(my_lock->name_);
     close(my_lock->fd_);
+
+    my_lock->fd_=-1;
```
Comment (bookshelfdave) [inline util/env_posix.cc:666]: setting the fd to -1 before checking for `close` errors

File: util/env_posix.cc (hunk 1 above)
Comment (matthewvon) [inline util/env_posix.cc:338]: There used to be a syslog about the return value back when I was looking for the failure.  Removed the syslog() call but left the ret_val= as convenience … if I was wrong about having fixed the problem.

File: util/env_posix.cc (hunk 2 above)
Comment (matthewvon) [inline util/env_posix.cc:405]: There used to be a syslog about the return value back when I was looking for the failure.  Removed the syslog() call but left the ret_val= as convenience … if I was wrong about having fixed the problem.

File: util/env_posix.cc (hunk 3 above)
Comment (matthewvon) [inline util/env_posix.cc:666]: Was more worried about double closes, not close errors.  None of the original google code verifies close errors … and dumping them into syslog is operationally useless.

**Comment by bookshelfdave** [comment]
+1

**Comment by bookshelfdave** [comment]
also, `make check` shows all tests passing

**Comment by matthewvon** [comment]
This was manually merged to the 1.4 branch.

**Pull request closed without merging**